staging/src/k8s.io/apiserver/pkg/endpoints/request/context.go

```diff
@@ -94,3 +97,18 @@ func AuditEventFrom(ctx context.Context) *audit.Event {
 	ev, _ := ctx.Value(auditKey).(*audit.Event)
 	return ev
 }
+
+func WithAuditAnnotations(parent context.Context, annotations map[string]string) context.Context {
```

Comment:
Consider inverting these so that callers that want to set audit annotations call something like:
```
func WithAuditAnnotation(parent context.Context, key, value string) context.Context
```
Used by callers setting an annotation.

Reply:
🤔 Hmm.

How about:

`func WithEmptyAuditAnnotations(parent context.Context) context.Context` which would just set up an empty map when called via
`func WithEmptyAuditAnnotations(handler http.Handler, sink audit.Sink, policy policy.Checker) http.Handler`

`func AddAuditAnnotation(ctx context.Context, key, value string)` for use by callers wanting to set audit annotations.  Note that this cannot allocate the map since it does not return a context.

Maybe a single `func AddEarlyAuditAnnotation(parent context.Context, key, value string) context.Context` would be enough if we are willing to assume that it will always be invoked in a context where it can overwrite the context (while it would always allocate a map even when auditing is disabled, it could at least only allocate one map if written correctly).

Reply:
https://github.com/kubernetes/kubernetes/compare/master...a6b605a0b08b6c196a83ad4ae0cb876713d09290 is the diff of the first approach and the current PR is up to date with the second approach ([delta](https://github.com/kubernetes/kubernetes/compare/49bd2157dff9c680e18fcfbda1f366355de6f174..cc8675c0607f4127f31c03c0d94ad2931d426fad)).

I think the second approach may be cleaner.